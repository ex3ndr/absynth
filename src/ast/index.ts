export { AbsynthLexer } from './lexer';
export { AbsynthParser } from './parser';

// export type ASTNodeType = 'number' | 'string' | 'declaration' | 'assignment' | 'program' | 'operation' | 'statements';

interface ASTBaseNode {
    first_column: number;
    first_line: number;
    last_column: number;
    last_line: number;
}

export interface ASTNumber extends ASTBaseNode {
    type: 'number';
    value: number;
}

export interface ASTString extends ASTBaseNode {
    type: 'string';
    value: string;
}

export interface ASTDeclaration extends ASTBaseNode {
    type: 'declaration';
    name: string;
}

export interface ASTReference extends ASTBaseNode {
    type: 'reference';
    name: string;
}

export interface ASTAssignment extends ASTBaseNode {
    type: 'assignment';
    left: ASTDeclaration;
    right: ASTExpression;
}

export interface ASTProgram extends ASTBaseNode {
    type: 'program';
    value: ASTNode[];
}

export interface ASTOperation extends ASTBaseNode {
    type: 'operation';
    op: string;
    left: ASTExpression;
    right: ASTExpression;
}

export interface ASTStatements extends ASTBaseNode {
    type: 'statements';
    statements: ASTNode[];
}

export interface ASTCall extends ASTBaseNode {
    type: 'call';
    name: string;
    arguments: ASTExpression[];
}

export interface ASTModel extends ASTBaseNode {
    type: 'model';
    name: ASTString;
    modifiers: ASTModifier[];
    statements: ASTStatements;
}

export interface ASTModifier extends ASTBaseNode {
    type: 'modifier';
    name: string;
}

export type ASTExpression = ASTOperation | ASTNumber | ASTString | ASTReference;

export type ASTNode = ASTNumber | ASTString | ASTDeclaration | ASTAssignment | ASTProgram | ASTOperation | ASTStatements | ASTCall | ASTModel | ASTModifier;