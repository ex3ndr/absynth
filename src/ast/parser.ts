import { AbsynthLexer, lexerTable } from './lexer';
import { Parser } from 'jison-gho';
import Lexer from 'lex';
import { ASTProgram } from '.';

interface ParserException {
    hash: {
        recoverable: boolean;
        lexer: Lexer;
        loc?: {
            first_line: number,
            first_column: number,
            last_line: number,
            last_column: number,
            range: number[]
        }
    };
}

function location(start?: number | string, end?: number | string) {
    let s = start || 1;
    let e = end || 1;
    return ', ' +
        'first_column: @' + s + '.first_column, ' +
        'first_line: @' + s + '.first_line, ' +
        'last_column: @' + e + '.last_column, ' +
        'last_line: @' + e + '.last_line';
}

export class AbsynthParser {
    private lexer: AbsynthLexer;
    private parser: Parser;

    constructor() {
        this.lexer = new AbsynthLexer();
        this.parser = new Parser({
            lex: {
                rules: lexerTable.map((v) => [v[0], v[1] !== '' ? 'return \'' + v[1] + '\';' : '']),
            },
            ebnf: {

                //
                // Program Root
                //

                'program': [['root_declarations EOF', 'return {type: \'program\', value: $1' + location() + '};']],

                //
                // Root Sections
                //

                'root_declarations': [
                    ['root_declarations root_declaration', '$$ = [...$1, $2]'],
                    ['root_declaration', '$$ = [$1]']
                ],

                'root_declaration': [
                    ['model', '$$ = $1'],
                    ['statements_block', '$$ = $1'],
                    ['statement', '$$ = $1'],
                ],

                //
                // Model Definitions
                //

                'model': [
                    ['modifiers T_MODEL id model_fields_block', '$$ = {type:\'model\', modifiers: $1, name: $3, fields: $4' + location(2, 4) + '};']
                ],

                'model_fields_block': [
                    ['T_BRACE_OPEN model_fields T_BRACE_CLOSE', '$$ = $2'],
                    ['T_BRACE_OPEN T_BRACE_CLOSE', '$$ = [];'],
                    //   ['model_block model_field'],
                    //   ['model_field']
                ],

                'model_fields': [
                    ['model_fields model_field', '$$ = [...$1, $2];'],
                    ['model_field', '$$ = [$1];']
                ],

                'model_field': [
                    ['modifiers T_FIELD id T_COLON type', '$$ = {type:\'field\', modifiers: $1, name: $3, field_type: $5' + location(2, 5) + '}']
                ],

                'modifiers': [
                    ['modifiers modifier', '$$ = [...$1, $2]'],
                    // ['T_MODIFIER', '$$ = {type:\'modifiers\', keywords: yytext' + location(1) + '}'],
                    ['', '$$ = []']
                ],

                'modifier': [
                    ['T_MODIFIER', '$$ = {type:\'modifier\', name: yytext' + location(1) + '}']
                ],

                'type': [
                    ['T_PRIMITIVE', '$$ = {type:\'type_primitive\', required: false, name: yytext' + location(1) + '}'],
                    ['T_PRIMITIVE T_EXCLAMATION', '$$ = {type:\'type_primitive\', required: true, name: $1' + location(1) + '}'],
                    ['id', '$$ = {type:\'type_reference\', required: false, id: $1' + location(1) + '}'],
                    ['id T_EXCLAMATION', '$$ = {type:\'type_reference\', required: true, id: $1' + location(1) + '}']
                ],

                //
                // Statements
                //

                'statements_block': [
                    ['T_BRACE_OPEN statements T_BRACE_CLOSE', '$$ = $2'],
                    ['T_BRACE_OPEN T_BRACE_CLOSE', '$$ = {type:\'statements\', statements: []' + location(1, 2) + '};'],
                ],

                'statements': [
                    ['statements statement', '$$ = {type:\'statements\', statements: [...$1.statements, $2]' + location(1, 2) + '};'],
                    ['statement', '$$ = {type:\'statements\', statements: [$1]' + location(1) + '};'],
                ],

                'statement': [
                    ['call T_SEMICOLON?', '$$ = $1;'],
                    ['assignment T_SEMICOLON?', '$$ = $1;'],
                    ['T_RETURN expression T_SEMICOLON?', '$$ = {type:\'return\', expression: $2' + location(1, 2) + ' }'],
                    ['declaration_enum T_SEMICOLON?', '$$ = $1;']
                ],

                //
                // Variable declaration
                //

                'declaration_variable': [['T_LET IDENTIFIER', '$$ = {type: \'declaration\', name: yytext ' + location(1) + '};']],
                'assignment': [['declaration_variable EQUALS expression', '$$ = {type: \'assignment\', left: $1, right: $3' + location(1, 3) + '};']],

                //
                // Enum
                //

                'declaration_enum': [['T_ENUM id EQUALS (constant_string[v] (T_BAR constant_string[v])*)[values]', '$$ = {type: \'enum\', name: $2, values: [$values[0], ...$values[1].map((v)=>v[1])]' + location(1, 'values') + '};']],

                //
                // Global function call
                //

                'call': [
                    ['IDENTIFIER T_BRACKET_OPEN arguments T_BRACKET_CLOSE', '$$ = {type:\'call\', name: $1, arguments: $3' + location(1, 4) + '};'],
                ],
                'arguments': [
                    ['expression T_COMMA arguments', '$$ = [$1, ...$3];'],
                    ['expression', '$$ = [$1];'],
                    ['', '$$ = [];']
                ],

                //
                // Expressions of all kind. Anything that eventually resolved to a value.
                //

                'expression': [
                    ['expression_additive', '$$ = $1;'],
                ],
                'expression_additive': [
                    ['expression_multiple T_PLUS expression_additive', '$$ = {type:\'operation\', op: \'+\', left: $1, right: $3' + location(1, 3) + '};'],
                    ['expression_multiple T_MINUS expression_additive', '$$ = {type:\'operation\', op: \'-\', left: $1, right: $3' + location(1, 3) + '};'],
                    ['expression_multiple', '$$ = $1;']
                ],
                'expression_multiple': [
                    ['constant T_MULT expression_multiple', '$$ = {type:\'operation\', op: \'*\', left: $1, right: $3' + location(1, 3) + '};'],
                    ['constant T_DIV expression_multiple', '$$ = {type:\'operation\', op: \'/\', left: $1, right: $3' + location(1, 3) + '};'],
                    ['expression_terminal', '$$ = $1;']
                ],
                'expression_terminal': [
                    ['call', '$$=$1'],
                    ['constant', '$$ = $1']
                ],
                'constant': [
                    ['IDENTIFIER', '$$ = {type: \'reference\', name: yytext' + location(1) + '};'],
                    ['constant_number', '$$ = $1;'],
                    ['constant_string', '$$ = $1;'],
                ],
                'id': [
                    ['IDENTIFIER', '$$ = {type: \'id\', name: yytext' + location(1) + '};'],
                ],
                'constant_number': [
                    ['NUMBER_HEX', '$$ = {type: \'number\', value: parseInt(yytext)' + location(1) + '};'],
                    ['NUMBER', '$$ = {type: \'number\', value: parseInt(yytext)' + location(1) + '};']
                ],
                'constant_string': [
                    ['STRING', '$$ = {type: \'string\', value: yytext.substring(1, yytext.length-1)' + location(1) + '};']
                ]
            }
        });
        // this.parser.lexer = this.lexer.lexerInstance();
    }

    parse(source: string): ASTProgram {
        return this.parser.parse(source);
    }

    parseDiagnostics(source: string): { loc: number[], msg: string } | undefined {
        try {
            this.parser.parse(source);
        } catch (e) {
            // console.warn(e);
            let ex = e as ParserException;
            if (ex.hash.loc) {
                console.log('Lexer error');
                console.log(ex.hash.loc);
                // console.log(ex.hash);
                return { loc: ex.hash.loc.range, msg: e.message };
            } else {
                // console.log((ex as any).hash);
            }
            // console.warn(ex.hash.lexer.index);
            // console.warn(ex.hash.lexer !== this.lexer.lexerInstance());
            // return ex.hash.lexer.index;
        }
        return undefined;
    }
}