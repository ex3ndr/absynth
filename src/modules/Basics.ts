import { Module, ModuleContext, location, ModuleOverlord, forward } from './_Base';

export class Basics extends Module {
    static ID = 'basics';

    private rootDeclarations = new Set<string>();

    getModuleId() {
        return Basics.ID;
    }

    addRootDeclaration(rule: string) {
        this.rootDeclarations.add(rule);
    }

    prepare(context: ModuleOverlord) {
        // Do nothing
    }

    applyContext(context: ModuleContext) {

        // Spaces And Comments
        context.registerLexeme('\\s+', '', 1); // Ignore spaces
        context.registerLexeme('\\/\\/[^\\n]+', '', 1); // Ignore single line comments
        context.registerLexeme('\\#[^\\n]+', '', 1); // Ignore single line comments
        context.registerLexeme('\\\\*.+\\*\\/', '', 1); // Ignore multiline comments
        context.registerLexeme('@@absynth[^\\n]+', '', 1); // Ignore special

        // Numeric constants
        context.registerLexeme('0x[a-fA-F\\d]+', 'NUMBER_HEX');
        context.registerLexeme('\\d+', 'NUMBER');
        context.registerLexeme('\\\'(\\.|[^\\\'])*\\\'', 'STRING');
        context.registerLexeme('true', 'TRUE');
        context.registerLexeme('false', 'FALSE');

        // Primitive types keywords
        context.registerLexeme('(string|int|float|bool)', 'T_PRIMITIVE');

        // Basic Punctuation
        context.registerLexeme(';', ';');
        context.registerLexeme(':', ':');
        context.registerLexeme('\\,', ',');
        context.registerLexeme('\\(', '(');
        context.registerLexeme('\\)', ')');
        context.registerLexeme('{', '{');
        context.registerLexeme('}', '}');
        context.registerLexeme('\\!', '!');
        context.registerLexeme('\\?', '?');
        context.registerLexeme('\\|', '|');

        // Arithmetic Operations
        context.registerLexeme('\\+', '+');
        context.registerLexeme('\\-', '-');
        context.registerLexeme('\\/', '/');
        context.registerLexeme('\\*', '*');
        context.registerLexeme('==', '==');
        context.registerLexeme('=', '=');

        // Identifier matcher
        context.registerLexeme('[a-zA-Z_][a-zA-Z_0-9]*', 'IDENTIFIER', 10); // Lowest priority to avoid matching before keywords

        // Default rules
        context.registerRules({
            'program': [['root_declarations EOF', 'return {type: \'program\', value: $1, ' + location() + '};']],
            'root_declarations': [
                ['root_declarations root_declaration', '$$ = [...$1, $2]'],
                ['root_declaration', '$$ = [$1]'],
            ],
            'root_declaration': [
                ...Array.from(this.rootDeclarations).map((v) => [v, forward()])
            ],
            'id': [
                ['IDENTIFIER', '$$ = {type: \'id\', name: yytext, ' + location(1) + '};'],
            ],
            'constant_number': [
                ['NUMBER_HEX', '$$ = {type: \'number\', value: parseInt(yytext), ' + location(1) + '};'],
                ['NUMBER', '$$ = {type: \'number\', value: parseInt(yytext), ' + location(1) + '};']
            ],
            'constant_string': [
                ['STRING', '$$ = {type: \'string\', value: yytext.substring(1, yytext.length-1), ' + location(1) + '};']
            ],
            'constant_boolean': [
                ['TRUE', '$$ = {type: \'boolean\', value: true, ' + location() + '}'],
                ['FALSE', '$$ = {type: \'boolean\', value: false, ' + location() + '}']
            ],
        });
    }
}