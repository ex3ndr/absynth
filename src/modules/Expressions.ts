import { Module, ModuleContext, forward, location, ModuleOverlord } from './_Base';

export class Expressions extends Module {
    static ID = 'expressions';

    private anomicExpressions = new Set<string>();

    getModuleId() {
        return Expressions.ID;
    }

    registerAtomicExpression(expression: string) {
        this.anomicExpressions.add(expression);
    }

    prepare(context: ModuleOverlord) {
        // Do nothing
    }

    applyContext(context: ModuleContext) {
        // Grammar Rules
        context.registerRules({
            'expression': [
                ['expression_additive', forward()],
            ],
            'expression_additive': [
                ['expression_multiple', forward()],
                ['expression_additive "+" expression_multiple', '$$ = {type:\'operation\', op: \'+\', left: $1, right: $3, ' + location(1, 3) + '};'],
                ['expression_additive "-" expression_multiple', '$$ = {type:\'operation\', op: \'-\', left: $1, right: $3, ' + location(1, 3) + '};'],
            ],
            'expression_multiple': [
                ['expression_atomic', forward()],
                ['expression_multiple "*" expression_atomic', '$$ = {type:\'operation\', op: \'*\', left: $1, right: $3, ' + location(1, 3) + '};'],
                ['expression_multiple "/" expression_atomic', '$$ = {type:\'operation\', op: \'/\', left: $1, right: $3, ' + location(1, 3) + '};'],
            ],
            'expression_atomic': [
                ['"(" expression ")"', forward(2)],
                ['id', forward()],
                ['constant_number', forward()],
                ['constant_string', forward()],
                ['constant_boolean', forward()]
            ],
        });
    }
}