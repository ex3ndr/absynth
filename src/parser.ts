import { AbsynthLexer } from './lexer';
import { Parser } from 'jison-gho';
import Lexer from 'lex';

export interface ASTNode {
    type: string;
}

export interface ASTNumber extends ASTNode {
    type: 'number';
    value: number;
}

export interface ASTString extends ASTNode {
    type: 'string';
    value: string;
}

export interface ASTDeclaration extends ASTNode {
    type: 'declaration';
    name: string;
}

export interface ASTAssignment extends ASTNode {
    type: 'assignment';
    left: ASTDeclaration;
    right: ASTNode;
}

export interface ASTProgram extends ASTNode {
    type: 'program';
    value: ASTNode;
}

export interface ASTOperation extends ASTNode {
    type: 'operation';
    op: 'string';
    left: ASTNode;
    right: ASTNode;
}

interface ParserException {
    hash: {
        recoverable: boolean;
        lexer: Lexer;
    };
}
export class AbsynthParser {
    private lexer: AbsynthLexer;
    private parser: Parser;

    constructor() {
        this.lexer = new AbsynthLexer();
        this.parser = new Parser({
            bnf: {
                'program': [['sections EOF', 'return {type: \'program\', value: $1};']],

                'sections': [
                    ['model', '$$=$1'],
                    ['block', '$$=$1'],
                    ['statements', '$$ = $1']
                ],

                'model': [
                    ['T_KEYWORD constant_string block', '$$ = {type:\'model\', name: $2, statements: $3};']
                ],

                'block': [
                    ['T_BRACE_OPEN statements T_BRACE_CLOSE', '$$ = $2'],
                    ['T_BRACE_OPEN T_BRACE_CLOSE', '$$ = {type:\'statements\', statements: []};'],
                    // ['statements', '$$ = $1']
                ],

                'statements': [
                    ['statements statement', '$$ = {type:\'statements\', statements: [...$1.statements, $2]};'],
                    ['statement', '$$ = {type:\'statements\', statements: [$1]};'],
                    // ['', '$$ = {type:\'statements\', statements: []};']
                ],

                'statement': [
                    ['call', '$$ = $1;'],
                    ['call T_SEMICOLON', '$$ = $1;'],
                    ['assignment', '$$ = $1;'],
                    ['assignment T_SEMICOLON', '$$ = $1;']
                ],

                'call': [
                    ['IDENTIFIER T_BRACKET_OPEN arguments T_BRACKET_CLOSE', '$$ = {type:\'call\', name: $1, arguments: $3};'],
                ],

                'arguments': [
                    ['expression T_COMMA arguments', '$$ = [$1, ...$3];'],
                    ['expression', '$$ = [$1];'],
                    ['', '$$ = [];']
                ],

                'assignment': [['declaration EQUALS expression', '$$ = {type: \'assignment\', left: $1, right: $3};']],
                'declaration': [['T_LET IDENTIFIER', '$$ = {type: \'declaration\', name: yytext};']],

                'expression': [
                    ['expression_additive', '$$ = $1;']
                ],
                'expression_additive': [
                    ['expression_multiple T_PLUS expression_additive', '$$ = {type:\'operation\', op: \'+\', left: $1, right: $3};'],
                    ['expression_multiple T_MINUS expression_additive', '$$ = {type:\'operation\', op: \'-\', left: $1, right: $3};'],
                    ['expression_multiple', '$$ = $1;']
                ],
                'expression_multiple': [
                    ['constant T_MULT expression_multiple', '$$ = {type:\'operation\', op: \'*\', left: $1, right: $3};'],
                    ['constant T_DIV expression_multiple', '$$ = {type:\'operation\', op: \'/\', left: $1, right: $3};'],
                    ['constant', '$$ = $1;']
                ],
                'constant': [
                    ['IDENTIFIER', '$$ = {type: \'reference\', name: yytext};'],
                    ['constant_number', '$$ = $1;'],
                    ['constant_string', '$$ = $1;'],
                ],
                'constant_number': [
                    ['NUMBER_HEX', '$$ = {type: \'number\', value: parseInt(yytext)};'],
                    ['NUMBER', '$$ = {type: \'number\', value: parseInt(yytext)};']
                ],
                'constant_string': [
                    ['STRING', '$$ = {type: \'string\', value: yytext};']
                ]
            }
        });
        this.parser.lexer = this.lexer.lexerInstance();
    }

    parse(source: string): ASTProgram {
        return this.parser.parse(source);
    }

    parseDiagnostics(source: string): number {
        try {
            this.parser.parse(source);
        } catch (e) {
            // console.warn(e);
            let ex = e as ParserException;
            // console.warn(ex.hash.lexer.index);
            // console.warn(ex.hash.lexer !== this.lexer.lexerInstance());
            return ex.hash.lexer.index;
        }
        return -1;
    }
}