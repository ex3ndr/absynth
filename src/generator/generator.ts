import { ASTProgram, ASTNode, ASTExpression, ASTModel, ASTType } from '../ast';
import { AbsynthScope } from './scoping';

function generateExpression(src: ASTExpression): string {
    if (src.type === 'string') {
        return '\'' + src.value + '\'';
    } else if (src.type === 'number') {
        return '' + src.value;
    } else if (src.type === 'reference') {
        return src.name;
    } else if (src.type === 'boolean') {
        return '' + src.value + '';
    } else if (src.type === 'operation') {
        if (src.op === '+') {
            return '(' + generateExpression(src.left) + ' + ' + generateExpression(src.right) + ')';
        } else if (src.op === '-') {
            return '(' + generateExpression(src.left) + ' - ' + generateExpression(src.right) + ')';
        } else if (src.op === '/') {
            return '(' + generateExpression(src.left) + ' / ' + generateExpression(src.right) + ')';
        } else if (src.op === '*') {
            return '(' + generateExpression(src.left) + ' * ' + generateExpression(src.right) + ')';
        } else {
            throw new GeneratorException('Unknown optration: ' + src.op, src);
        }
    } else if (src.type === 'call') {
        let actualName: string | undefined;
        if (src.name === 'println') {
            actualName = 'console.log';
        }
        if (!actualName) {
            throw new GeneratorException('Unable to resolve ' + src.name, src);
        }
        return actualName + '(' + src.arguments.map(generateExpression).join(', ') + ')';
    } else {
        throw new GeneratorException('Unknown expression' + src, src);
    }
}

function resolveModelKeyType(src: ASTModel): string {
    let idField = src.fields.find((v) => !!v.modifiers.find((v2) => v2.name === 'primary'));
    if (!idField) {
        throw new GeneratorException('Unable to find type', src);
    }
    return generateReferencedType(idField.field_type);
}

function resolveModelKeyName(src: ASTModel): string {
    let idField = src.fields.find((v) => !!v.modifiers.find((v2) => v2.name === 'primary'));
    if (!idField) {
        throw new GeneratorException('Unable to find type', src);
    }
    return idField.name.name;
}

function generateReferencedType(src: ASTType): string {
    if (src.type === 'type_primitive') {
        if (src.name === 'int') {
            return 'number';
        }
        if (src.name === 'string') {
            return 'string';
        }
    } else if (src.type === 'type_reference') {
        if (src.resolvedType!!.type === 'model') {
            return resolveModelKeyType(src.resolvedType!! as ASTModel);
        } else if (src.resolvedType.type === 'enum') {
            return src.resolvedType.name.name;
        }
    }
    throw new GeneratorException('Unable to resolve type', src);
}

function capitalize(src: string): string {
    return src.charAt(0).toUpperCase() + src.substring(1);
}

function generateModel(src: ASTModel): string {
    let pr = src.modifiers.find((v) => v.name === 'private');
    if (pr) {
        throw new GeneratorException('Model can\'t be private!', pr);
    }

    //
    // Attributes
    //

    let res = 'export interface ' + src.name.name + 'Attributes {';
    let ex = new Set<string>();
    for (let f of src.fields) {
        if (ex.has(f.name.name)) {
            throw new GeneratorException('Duplicate field declaration', f.name);
        }
        ex.add(f.name.name);

        let type = generateReferencedType(f.field_type);
        if (!f.field_type.required) {
            type += ' | null';
        }
        let name = f.name.name;
        if (f.field_type.type === 'type_reference') {
            if (f.field_type.resolvedType!!.type === 'model') {
                name = f.name.name + capitalize(resolveModelKeyName(f.field_type.resolvedType!! as ASTModel));
            }
        }
        res += '\n    ' + name + ': ' + type + ';';
    }
    res += '\n}\n\n';

    //
    // Instante
    //

    res += 'export interface ' + src.name.name + ' extends sequelize.Instance<Partial<' + src.name.name + 'Attributes>>, ' + src.name.name + 'Attributes {';
    for (let f of src.fields) {
        if (f.field_type.type === 'type_reference') {
            if (f.field_type.resolvedType!!.type === 'model') {
                let type = f.field_type.resolvedType!!.name.name;
                let name = f.name.name;
                if (!f.field_type.required) {
                    type += ' | null';
                }
                res += '\n    ' + name + ': ' + type + ' | undefined;';
            }
            // res += '\n    get' + capitalize(name) + '(options?: any): Promise<' + type + '>;';
        }
    }
    res += '\n    readonly createdAt: Date;';
    res += '\n    readonly updatedAt: Date;';
    for (let f of src.fields) {
        if (f.field_type.type === 'type_reference') {
            if (f.field_type.resolvedType!!.type === 'model') {
                let type = f.field_type.resolvedType!!.name.name;
                let name = f.name.name;
                if (!f.field_type.required) {
                    type += ' | null';
                }
                // res += '\n    ' + name + ': ' + type + ' | undefined;';
                res += '\n    get' + capitalize(name) + '(options?: any): Promise<' + type + '>;';
            }
        }
    }
    res += '\n}\n';

    //
    // Table
    //

    res += 'export const ' + src.name.name + 'Table = connection.define<' + src.name.name + ', Partial<' + src.name.name + 'Attributes>>(\'' + src.name.name + '\', {';

    for (let f of src.fields) {
        let attrs = 'allowNull: ' + !f.field_type.required;
        if (f.modifiers.find((v) => v.name === 'primary')) {
            attrs += ', primaryKey: true, autoIncrement: true';
        }
        if (f.field_type.type === 'type_primitive') {
            if (f.field_type.name === 'string') {
                attrs += ', type: sequelize.STRING';
            } else if (f.field_type.name === 'int') {
                attrs += ', type: sequelize.INTEGER';
            } else {
                throw new GeneratorException('Unknown type', f.field_type);
            }
            // let type = f.field_type.resolvedType!!.name.name;
            // let name = f.name.name;
            // if (!f.field_type.required) {
            //     type += ' | null';
            // }
            // // res += '\n    ' + name + ': ' + type + ' | undefined;';
            // res += '\n    get' + capitalize(name) + '(options?: any): Promise<' + type + '>;';
            res += '\n    ' + f.name.name + ': { ' + attrs + ' },';
        } else if (f.field_type.type === 'type_reference') {
            // Ignore
            // res += '\n    ' + f.name.name + ': { ' + attrs + ', },';
        } else {
            throw new GeneratorException('Unknown type', f.field_type);
        }
    }

    res += '\n});\n';

    for (let f of src.fields) {
        if (f.field_type.type === 'type_reference') {
            res += '\n' + src.name.name + 'Table.belongsTo(' + f.field_type.resolvedType!!.name.name + 'Table, { as: \'' + f.name.name + '\' });';
            //
        }
    }

    return res;
}

function generateStatement(src: ASTNode): string {
    if (src.type === 'call') {
        let actualName: string | undefined;
        if (src.name === 'println') {
            actualName = 'console.log';
        }
        if (!actualName) {
            throw new GeneratorException('Unable to resolve ' + src.name, src);
        }
        return actualName + '(' + src.arguments.map(generateExpression).join(', ') + ');';
    } else if (src.type === 'assignment') {
        return 'let ' + src.left.name + ' = ' + generateExpression(src.right) + ';';
    } else if (src.type === 'statements') {
        let res = '';
        for (let s of src.statements) {
            if (res.length > 0) {
                res += '\n';
            }
            res += generateStatement(s);
        }
        return '{\n' + res + '\n}';
    } else if (src.type === 'model') {
        return generateModel(src);
    } else if (src.type === 'enum') {
        return 'type ' + src.name.name + ' = ' + src.values.map((v) => '\'' + v.value + '\'').join(' | ') + ';';
    } else {
        console.log(JSON.stringify(src));
        throw new GeneratorException('Unsupported mode', src);
    }
}

function generateProgram(src: ASTProgram): string {
    let res = '';
    for (let s of src.value) {
        if (res.length > 0) {
            res += '\n';
        }
        res += generateStatement(s);
    }
    return res;
}

export class GeneratorException extends Error {
    node: ASTNode;
    constructor(message: string, node: ASTNode) {
        super(message);
        this.node = node;
    }
}

export class AbsynthGenerator {

    generate(src: ASTProgram): string {
        let scope = new AbsynthScope();
        scope.collectDeclarations(src);
        scope.resolveReferences(src);
        return generateProgram(src);
    }
}