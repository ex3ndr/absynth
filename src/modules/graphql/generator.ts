import { ASTProgram, ASTNode } from "../../parser";
import { GeneratorException } from "../core/GeneratorException";

interface GraphqlType {
    type: 'interface' | 'input' | 'type';
    name: string;
    fields: { name: string, type: GraphqlTypeReference }[];
}

interface GraphqlTypeReference {
    type: 'array' | 'reference';
    name: string;
    required: boolean;
}

class ASTResolver {
    types = new Map<String, GraphqlType>();
}

function buildAstNode(src: ASTNode, resolver: ASTResolver) {
    if (src.type === 'type') {
        let name = src.name.name;
        console.log(name);
        if (resolver.types.has(name)) {
            throw new GeneratorException('Double declaration', src);
        }
        resolver.types.set(name, {
            type: 'type',
            name: name,
            fields: []
        });
        // src.name.name;
        // return 'type ' + src.name.name;
    } else {
        throw new GeneratorException('Unsupported mode', src);
    }
}

function buildAst(src: ASTProgram) {
    let resolver = new ASTResolver();
    for (let s of src.value) {
        // console.log(s);
        buildAstNode(s, resolver);
    }
    return resolver;
}

// function generateNode(src: ASTNode): string {
//     if (src.type === 'type') {
//         src.name.name;
//         return 'type ' + src.name.name;
//     } else {
//         throw new GeneratorException('Unsupported mode', src);
//     }
// }

export function generateProgram(src: ASTProgram): string {
    return JSON.stringify(buildAst(src).types);
    // return src.value.map((v) => generateNode(v)).join('\n');
}