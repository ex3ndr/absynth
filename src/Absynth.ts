import { ModuleOverlord, Module } from "./modules/_Base";
import { AbsynthLexer, AbsynthParser } from "./ast";

export class Absynth {
    readonly lexer: AbsynthLexer;
    readonly parser: AbsynthParser;
    private overlord: ModuleOverlord;

    static create(src: string, modules: Module[]) {
        if (!src.startsWith('@@absynth')) {
            throw Error('File is not start with @@absynth keyword');
        }

        let declaration = src.split('\n')[0].trim();
        declaration = declaration.substring('@@absynth'.length + 1);
        declaration = declaration.substring(0, declaration.length - 1);
        let mods = declaration.split(',').map((v) => v.trim());
        let m = modules.filter((v) => mods.indexOf(v.getModuleId()) >= 0);
        return new Absynth(m);
    }

    constructor(modules: Module[]) {
        this.overlord = new ModuleOverlord();
        for (let m of modules) {
            this.overlord.addModule(m);
        }
        let config = this.overlord.prepare();
        this.lexer = new AbsynthLexer(config.lex);
        this.parser = new AbsynthParser(config.lex, config.ebnf);
    }
}