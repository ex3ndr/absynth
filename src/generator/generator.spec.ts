import { AbsynthParser } from '../ast/parser';
import { AbsynthGenerator } from './generator';
import { Absynth } from '../Absynth';
import { Basics } from '../modules/Basics';
import { Expressions } from '../modules/Expressions';
import { Experiments } from '../modules/Experiments';
function createParser() {
    return new Absynth([new Basics(), new Expressions(), new Experiments()]).parser;
}
describe('Generator', () => {
    it('should generate simple statements', () => {
        let parser = createParser();
        let generator = new AbsynthGenerator();
        let ast = parser.parse(`
            println('Hello' + ' ' + 'World!')
            println(123 + 6/2)
        `);
        let app = generator.generate(ast);
        expect(app).toMatchSnapshot();
    });
});