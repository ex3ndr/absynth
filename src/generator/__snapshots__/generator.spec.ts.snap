// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator should generate simple statements 1`] = `
"console.log((('Hello' + ' ') + 'World!'));
console.log((123 + (6 / 2)));"
`;
